pkg/agent/manager/cache/cache.go

```diff
@@ -0,0 +1,175 @@
+	Entry(regEntry *common.RegistrationEntry) *Entry
+	// SetEntry puts a new cache entry at the end of the list of entries for the entry's RegistrationEntry.
+	SetEntry(entry *Entry)
+	// DeleteEntries deletes all the Entries for the specified RegistrationEntry, returns an integer
+	// with the number of cache entries that were removed.
+	DeleteEntries(regEntry *common.RegistrationEntry) int
+	// DeleteEntry removes the first cache entry for the specified RegistrationEntry if any,
+	// returns true if it removed some entry or false otherwise.
+	DeleteEntry(regEntry *common.RegistrationEntry) bool
+	// Entries returns all the in force cached entries.
+	Entries() chan Entry
+	// IsEmpty returns true if this cache doesn't have any entry.
+	IsEmpty() bool
+	// Register a Subscriber and return WorkloadUpdate on the subscriber's channel
+	Subscribe(sub *Subscriber)
+}
+
+type cacheImpl struct {
+	// Map keyed by a combination of SpiffeId + ParentId + Selectors holding a list of
+	// Entry instances ordered by SVID expiration date.
+	cache       map[string][]Entry
+	log         logrus.FieldLogger
+	m           sync.Mutex
+	Subscribers subscribers
+	bundle      []*x509.Certificate
+}
+
+// New creates a new Cache.
+func New(log logrus.FieldLogger, bundle []*x509.Certificate) Cache {
+	return &cacheImpl{
+		cache:  make(map[string][]Entry),
+		log:    log.WithField("subsystem_name", "cache"),
+		bundle: bundle,
+	}
+}
+
+func (c *cacheImpl) SetServerBundle(bundle []*x509.Certificate) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.bundle = bundle
+}
+
+func (c *cacheImpl) serverBundle() []*x509.Certificate {
+	c.m.Lock()
+	defer c.m.Unlock()
+	return c.bundle
+}
+
+func (c *cacheImpl) Entries() chan Entry {
+	c.m.Lock()
+	defer c.m.Unlock()
+	entries := make(chan Entry, len(c.cache))
+	for _, e := range c.cache {
+		// Only return the first element for each array of entries because it is the
+		// in force entry.
+		entries <- e[0]
+	}
+	close(entries)
+	return entries
+}
+
+func (c *cacheImpl) Subscribe(sub *Subscriber) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	var entries []Entry
+	for _, e := range c.cache {
+		regEntrySelectors := selector.NewSetFromRaw(e[0].RegistrationEntry.Selectors)
+		if selector.NewSetFromRaw(sub.sel).IncludesSet(regEntrySelectors) {
+			entries = append(entries, e[0])
+		}
+	}
+	select {
+	case sub.C <- &WorkloadUpdate{cacheEntries: entries, bundle: c.bundle}:
+	default:
+		c.log.Warn("Failed to send workload update sid:%v method:subscribe", sub.sid)
+
+	}
+	c.Subscribers.Add(sub)
+}
+
+func (c *cacheImpl) Entry(regEntry *common.RegistrationEntry) *Entry {
+	key := util.DeriveRegEntryhash(regEntry)
+	c.m.Lock()
+	defer c.m.Unlock()
+	if entries, found := c.cache[key]; found {
+		return &entries[0]
+	}
+	return nil
+}
+
+func (c *cacheImpl) SetEntry(entry *Entry) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	subs := c.Subscribers.Get(entry.RegistrationEntry.Selectors)
+	for _, sub := range subs {
```

Comment:
Should we pack this into a dedicated function? We do it a couple different times - it's also a little hard to read IMO